### PR TITLE
simplifying native example code and creating feed and widget 4X1 screens

### PR DESCRIPTION
**Pull request by voltfmanSlava**

### CODE REVIEW

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+ */
+public class FeedFragment extends Fragment {
+    private RecyclerView mRecyclerView;
+    private ArrayList<Object> mData = new ArrayList<>();
+    private TBLNativeUnit mNativeUnit = null;
+
+    @Nullable
+    @Override
+    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
+        View root = inflater.inflate(R.layout.fragment_native_feed, container, false);
```
Comment (EitanSchwartz) [inline Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java:43]: Maybe add a comment: "Define a RecyclerView in the XML layout file"

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+import com.taboola.android.tblnative.TBLNativeUnit;
+import com.taboola.android.tblnative.TBLPlacement;
+import com.taboola.android.tblnative.TBLRecommendationRequestCallback;
+import com.taboola.android.tblnative.TBLRecommendationsResponse;
+import com.taboola.android.tblnative.TBLRequestData;
+import com.taboola.sdk3example.R;
+
+import java.util.ArrayList;
+
+import static java.sql.DriverManager.println;
```
Comment (EitanSchwartz) [inline Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java]: Why are we doing 'println' everywhere?

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+     */
+    private void setupRecyclerView(Context context) {
+        mRecyclerView.setLayoutManager(new LinearLayoutManager(context));
+
+        // Load more Taboola content when reaching scroll bottom
+        mRecyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
+            @Override
+            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
+                super.onScrollStateChanged(recyclerView, newState);
+                if (!recyclerView.canScrollVertically(1) && newState == RecyclerView.SCROLL_STATE_IDLE) {
```
Comment (EitanSchwartz) [inline Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java:70]: Worth explaining that this is how we're loading more content when the user saw all items

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+            });
+        }
+    }
+
+    /**
+     * Taboola "Native Integration" currently differentiates between first and subsequent content fetch calls.
+     * This method fetches additional items for this Feed implementation.
+     */
+    private void loadNextRecommendationsBatch() {
+        mNativeUnit.getNextRecommendationsBatch(null, new TBLRecommendationRequestCallback() {
```
Comment (EitanSchwartz) [inline Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java]: if (mNativeUnit != null)

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+                println(String.format("Taboola | onRecommendationsFailed: %s", throwable.getLocalizedMessage()));
+            }
+        });
+    }
+
+    /**
+     * This method parses the response with its items and adds them to the RecyclerView Adapter.
+     */
+    private void addRecommendationToFeed(TBLRecommendationsResponse recommendationsResponse) {
+        TBLPlacement placement = recommendationsResponse.getPlacementsMap().values().iterator().next();
```
Comment (EitanSchwartz) [inline Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java]: if (recommendationsResponse !=null)

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    /**
+     * Define a Page that represents this screen, get a Unit from it, add it to screen and fetch its content
+     * Notice: A Unit of unlimited items, called "Feed" in Taboola, can be set in TBL_PLACEMENT_TYPE.PAGE_BOTTOM only.
+     */
+    private void getTaboolaUnit() {
+        // Define a page to control all Unit placements on this screen
+        TBLNativePage nativePage = Taboola.getNativePage("text", "https://blog.taboola.com");
+
+        // Define a single Unit to display
+        mNativeUnit = nativePage.build("list_item", new TBLNativeListener() {
```
Comment (EitanSchwartz) [inline Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java]: I'm missing something here, the method signature for NativePage.build, in 3.x master, is:

`build(String placementName, @Nullable TBLPublisherInfo tblPublisherInfo, TBLNativeListener tblNativeListener, @Nullable TBLRequestData tblRequestData)`

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/NativeFeedAdapter.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
+        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
+        View linearLayout = inflater.inflate(R.layout.feed_taboola_item, parent, false);
+        return new TaboolaItemViewHolder(linearLayout);
+    }
+
+    @Override
+    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
+        if (mData == null) {
+            return;
```
Comment (EitanSchwartz) [inline Java/app/src/main/java/com/taboola/sdk3example/sdk_native/NativeFeedAdapter.java:41]: plz add a log msg

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/NativeFeedAdapter.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        }
+
+        TBLRecommendationItem item = (TBLRecommendationItem) mData.get(position);
```
Comment (EitanSchwartz) [inline Java/app/src/main/java/com/taboola/sdk3example/sdk_native/NativeFeedAdapter.java:44]: Can mData.get(position) return null?

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/NativeFeedAdapter.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        ViewGroup adContainer = ((TaboolaItemViewHolder) holder).mAdContainer;
+        Context context = adContainer.getContext();
```
Comment (EitanSchwartz) [inline Java/app/src/main/java/com/taboola/sdk3example/sdk_native/NativeFeedAdapter.java]: Worth explaining how Taboola  Native integration expects the publisher to implement their Taboola UI using the different Taboola views

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/NativeFeedAdapter.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+            return 0;
+        }
+
+        return mData.size();
+    }
+
+    @Override
+    public void onViewRecycled(@NonNull RecyclerView.ViewHolder holder) {
+        super.onViewRecycled(holder);
+        if (holder instanceof TaboolaItemViewHolder) {
```
Comment (EitanSchwartz) [inline Java/app/src/main/java/com/taboola/sdk3example/sdk_native/NativeFeedAdapter.java:86]: comment explaining that this is how we separate items that should contain Taboola

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java (hunk 1 above)
Comment (voltfmanSlava) [reply Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java:43]: Done

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java (hunk 3 above)
Comment (voltfmanSlava) [reply Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java:70]: Done

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java (hunk 4 above)
Comment (voltfmanSlava) [reply Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java]: Done

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java (hunk 5 above)
Comment (voltfmanSlava) [reply Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java]: Done

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java (hunk 6 above)
Comment (voltfmanSlava) [reply Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java]: The same code as in Kotlin project it takes TBLPublisherInfo from global

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java (hunk 2 above)
Comment (voltfmanSlava) [reply Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java]: Taken from kotlin project. changed to Log.d

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/NativeFeedAdapter.java (hunk 7 above)
Comment (voltfmanSlava) [reply Java/app/src/main/java/com/taboola/sdk3example/sdk_native/NativeFeedAdapter.java:41]: Done

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/NativeFeedAdapter.java (hunk 8 above)
Comment (voltfmanSlava) [reply Java/app/src/main/java/com/taboola/sdk3example/sdk_native/NativeFeedAdapter.java:44]: Done

Comment (voltfmanSlava) [reply Java/app/src/main/java/com/taboola/sdk3example/sdk_native/NativeFeedAdapter.java:44]: Done

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/SDKNativeMenu.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+import androidx.appcompat.app.AppCompatActivity;
+import androidx.appcompat.widget.Toolbar;
+import androidx.fragment.app.Fragment;
+import androidx.fragment.app.FragmentTransaction;
+
+import com.taboola.sdk3example.R;
+
+
+public class SDKNativeMenu extends AppCompatActivity   implements SDKNativeMenuFragment.OnFragmentInteractionListener {
```
Comment (EitanSchwartz) [inline Java/app/src/main/java/com/taboola/sdk3example/sdk_native/SDKNativeMenu.java]: Why is our Activity called a "..Menu"?

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/SDKNativeMenu.java [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+    protected void onCreate(Bundle savedInstanceState) {
+        requestWindowFeature(Window.FEATURE_NO_TITLE);
+        this.getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,WindowManager.LayoutParams.FLAG_FULLSCREEN);
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.sdk_native_menu);
+
+
+        mToolbar = findViewById(R.id.toolbar);
+        setSupportActionBar(mToolbar);
+        getSupportActionBar().setLogo(R.drawable.ic_taboola);
```
Comment (EitanSchwartz) [inline Java/app/src/main/java/com/taboola/sdk3example/sdk_native/SDKNativeMenu.java]: no need to getSupportActionBar twice, also, I think you can just use mToolBar

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/SDKNativeMenu.java [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+                }
+            } else {
+                if (getSupportActionBar() != null) {
+                    getSupportActionBar().setTitle(getSupportFragmentManager().getBackStackEntryAt(lastBackStackEntryCount).getName());
+                    showBackArrow(true);
+                }
+            }
+        });
+
+        getSupportFragmentManager().beginTransaction()
```
Comment (EitanSchwartz) [inline Java/app/src/main/java/com/taboola/sdk3example/sdk_native/SDKNativeMenu.java]: getSupportFragmentManager() only once

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/SDKNativeMenu.java [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+        getSupportFragmentManager().addOnBackStackChangedListener(() -> {
+            int lastBackStackEntryCount = getSupportFragmentManager().getBackStackEntryCount() - 1;
+
+            if (lastBackStackEntryCount < 0) {
+                resetToolbarTitle();
+                if (getSupportActionBar() != null) {
+                    showBackArrow(false);
+                }
+            } else {
+                if (getSupportActionBar() != null) {
```
Comment (EitanSchwartz) [inline Java/app/src/main/java/com/taboola/sdk3example/sdk_native/SDKNativeMenu.java]: again, use mToolBar

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/SDKNativeMenu.java [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+
+        getSupportFragmentManager().beginTransaction()
+                .replace(R.id.container, new SDKNativeMenuFragment()).commit();
+        prepareTaboolaLogoRotation();
+
+    }
+
+
+    private void showBackArrow(boolean shouldShowBackButton) {
+        if (getSupportActionBar() != null) {
```
Comment (EitanSchwartz) [inline Java/app/src/main/java/com/taboola/sdk3example/sdk_native/SDKNativeMenu.java]: getSupportActionBar() mtoolbar everywhere

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/WidgetFragment.java [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+
+    /**
+     * Define a Page that represents this screen, get a Unit from it, add it to screen and fetch its content
+     * Notice: A Unit of unlimited items, called "Feed" in Taboola, can be set in TBL_PLACEMENT_TYPE.PAGE_BOTTOM only.
+     */
+    private TBLNativeUnit getTaboolaUnit() {
+        // Define a page to control all Unit placements on this screen
+        TBLNativePage nativePage = Taboola.getNativePage("text", "https://blog.taboola.com");
+
+        return nativePage.build("list_item", new TBLNativeListener() {
```
Comment (EitanSchwartz) [inline Java/app/src/main/java/com/taboola/sdk3example/sdk_native/WidgetFragment.java]: removed build() signature, fix all places in pr plz

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/WidgetFragment.java [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+        return root;
+    }
+
+    /**
+     * Define a Page that represents this screen, get a Unit from it, add it to screen and fetch its content
+     * Notice: A Unit of unlimited items, called "Feed" in Taboola, can be set in TBL_PLACEMENT_TYPE.PAGE_BOTTOM only.
+     */
+    private TBLNativeUnit getTaboolaUnit() {
+        // Define a page to control all Unit placements on this screen
+        TBLNativePage nativePage = Taboola.getNativePage("text", "https://blog.taboola.com");
```
Comment (EitanSchwartz) [inline Java/app/src/main/java/com/taboola/sdk3example/sdk_native/WidgetFragment.java:66]: Please make sure all this code runs on 3.5.0

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/WidgetFragment.java [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+        if (placement != null && placement.getItems().size() > 0) {
+
+            // We asked for 4 item so we go over all items returned from Taboola
+            // NOTE: In this example, we search for the items relevant for the only placement in this page
+
+            for (TBLRecommendationItem item : placement.getItems()) {
+                Context context = contentLayout.getContext();
+                ViewGroup.LayoutParams layoutParams = new ViewGroup.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
+                TBLImageView thumbnailView = item.getThumbnailView(context);
+                TBLTextView titleView = item.getTitleView(context);
```
Comment (EitanSchwartz) [inline Java/app/src/main/java/com/taboola/sdk3example/sdk_native/WidgetFragment.java:105]: Maybe put all the little parts of the Native UI in a method, explaining this unique behavior of building UI by the publisher

**Review by EitanSchwartz — APPROVED** [review]

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java (hunk 6 above)
Comment (voltfmanSlava) [reply Java/app/src/main/java/com/taboola/sdk3example/sdk_native/FeedFragment.java]: Done

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/SDKNativeMenu.java (hunk 11 above)
Comment (voltfmanSlava) [reply Java/app/src/main/java/com/taboola/sdk3example/sdk_native/SDKNativeMenu.java]: as parity to SDKWebMenu and SDKClassicMenu those activities with menu to choose which screen to display. will change all of them to end with "Activity"

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/SDKNativeMenu.java (hunk 12 above)
Comment (voltfmanSlava) [reply Java/app/src/main/java/com/taboola/sdk3example/sdk_native/SDKNativeMenu.java]: Done

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/SDKNativeMenu.java (hunk 13 above)
Comment (voltfmanSlava) [reply Java/app/src/main/java/com/taboola/sdk3example/sdk_native/SDKNativeMenu.java]: Done

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/WidgetFragment.java (hunk 16 above)
Comment (voltfmanSlava) [reply Java/app/src/main/java/com/taboola/sdk3example/sdk_native/WidgetFragment.java]: Done

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/SDKNativeMenu.java (hunk 15 above)
Comment (voltfmanSlava) [reply Java/app/src/main/java/com/taboola/sdk3example/sdk_native/SDKNativeMenu.java]: Done

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/SDKNativeMenu.java (hunk 14 above)
Comment (voltfmanSlava) [reply Java/app/src/main/java/com/taboola/sdk3example/sdk_native/SDKNativeMenu.java]: Done

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/NativeFeedAdapter.java (hunk 10 above)
Comment (voltfmanSlava) [reply Java/app/src/main/java/com/taboola/sdk3example/sdk_native/NativeFeedAdapter.java:86]: Done

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/NativeFeedAdapter.java (hunk 9 above)
Comment (voltfmanSlava) [reply Java/app/src/main/java/com/taboola/sdk3example/sdk_native/NativeFeedAdapter.java]: Done

File: Java/app/src/main/java/com/taboola/sdk3example/sdk_native/WidgetFragment.java (hunk 18 above)
Comment (voltfmanSlava) [reply Java/app/src/main/java/com/taboola/sdk3example/sdk_native/WidgetFragment.java:105]: Done

**Review by EitanSchwartz — APPROVED** [review]

**Pull request merged**